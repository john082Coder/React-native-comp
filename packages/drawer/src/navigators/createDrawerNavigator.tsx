import * as React from 'react';
import { Dimensions, Platform, ScrollView, I18nManager } from 'react-native';
import { createNavigator, ThemeColors } from '@react-navigation/core';
import { SafeAreaView } from '@react-navigation/native';
import DrawerRouter from '../routers/DrawerRouter';
import DrawerView from '../views/DrawerView';
import DrawerItems, { Props } from '../views/DrawerNavigatorItems';

// A stack navigators props are the intersection between
// the base navigator props (navgiation, screenProps, etc)
// and the view's props

const defaultContentComponent = (props: Props) => (
  <ScrollView alwaysBounceVertical={false}>
    <SafeAreaView forceInset={{ top: 'always', horizontal: 'never' }}>
      <DrawerItems {...props} />
    </SafeAreaView>
  </ScrollView>
);

const DefaultDrawerConfig = {
  drawerWidth: () => {
    /*
     * Default drawer width is screen width - header height
     * with a max width of 280 on mobile and 320 on tablet
     * https://material.io/guidelines/patterns/navigation-drawer.html
     */
    const { height, width } = Dimensions.get('window');
    const smallerAxisSize = Math.min(height, width);
    const isLandscape = width > height;
    const isTablet = smallerAxisSize >= 600;
    const appBarHeight = Platform.OS === 'ios' ? (isLandscape ? 32 : 44) : 56;
    const maxWidth = isTablet ? 320 : 280;

    return Math.min(smallerAxisSize - appBarHeight, maxWidth);
  },
  contentComponent: defaultContentComponent,
  drawerPosition: I18nManager.isRTL ? 'right' : 'left',
  keyboardDismissMode: 'on-drag',
  drawerBackgroundColor: {
    light: ThemeColors.light.bodyContent,
    dark: ThemeColors.dark.bodyContent,
  },
  overlayColor: {
    light: 'rgba(0, 0, 0, 0.5)',
    dark: 'rgba(0, 0, 0, 0.5)',
  },
  drawerType: 'front',
  hideStatusBar: false,
  statusBarAnimation: 'slide',
};

const DrawerNavigator = (routeConfigs: object, config: any = {}) => {
  const mergedConfig = { ...DefaultDrawerConfig, ...config };
  const drawerRouter = DrawerRouter(routeConfigs, mergedConfig);
  const navigator = createNavigator(DrawerView, drawerRouter, mergedConfig);
  return navigator;
};

export default DrawerNavigator;
