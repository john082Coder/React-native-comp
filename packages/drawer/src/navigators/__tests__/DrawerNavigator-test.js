import React, { Component } from 'react';
import { View } from 'react-native';
import renderer from 'react-test-renderer';

import createDrawerNavigator from '../createDrawerNavigator';

class HomeScreen extends Component {
  static navigationOptions = ({ navigation }) => ({
    title: `Welcome ${
      navigation.state.params ? navigation.state.params.user : 'anonymous'
    }`,
    gesturesEnabled: true,
  });

  render() {
    return <View style={{ flex: 1 }} />;
  }
}

describe('createDrawerNavigator', () => {
  it('renders successfully', () => {
    const MyDrawerNavigator = createDrawerNavigator({ Home: HomeScreen });
    const rendered = renderer.create(<MyDrawerNavigator />).toJSON();

    expect(rendered).toMatchSnapshot();
  });
});
