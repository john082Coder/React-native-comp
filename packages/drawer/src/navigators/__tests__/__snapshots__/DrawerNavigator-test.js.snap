// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createDrawerNavigator renders successfully 1`] = `
<RCTView
  drawerBackgroundColor="white"
  drawerLockMode={undefined}
  drawerPosition="left"
  drawerWidth={320}
  onDrawerClose={[Function]}
  onDrawerOpen={[Function]}
  renderNavigationView={[Function]}
  useNativeAnimations={true}
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  />
</RCTView>
`;
